ykrt/src/deopt.rs

```diff
@@ -306,8 +306,8 @@ unsafe extern "C" fn __ykrt_deopt(
         if let Some(st) = guard.getct() {
             let registers = Registers::from_ptr(rsp);
             let live_vars = ctr.smap().get(&retaddr.try_into().unwrap()).unwrap();
-            let mut ykctrlpvars = Vec::new();
-            for (_i, locs) in live_vars.iter().skip(1).enumerate() {
+            let mut ykctrlpvars = vec![0; live_vars.len() - 1];
```

Comment:
Do we know that `live_vars.len()` is > 0?

Reply:
It would require there to be no live variables whatsoever, which, while highly unlikely, Im not confident enough about to categorically rule out. So I have two possible fixes for you:
```rust
            let mut ykctrlpvars = vec![0; live_vars.len()];
```
or
```rust
debug_assert!(live_vars.len() > 0);
```

Reply:
You could also use `checked_sub` instead of `-` and/or add a check in the trace compiler that traces always have 1 live variable or more.

[I'm actually not really sure why the `-1` is here though! If there are *n* live vars, why would we allocate space for *n - 1*?]

Reply:
Wouldn't a `checked_add` be less performant? The reason we do `n-1` is because we skip the first element of `live_vars` as it isn't needed. But we could just use `n` to size the vec. That extra element isn't going to kill us.

Reply:
If we skip the first variable then presumably we *do* know there are > 0 variables? What is that first variable?

Reply:
Unlike stackmaps generated by stackmap calls, stackmaps generated by the `llvm_deoptimise` intrinsic has an additional live variable at the beginning containing 3 values CC, Flags, and NumDeopts. These values are of no interest to us so we just skip them here. What I'm not certain about is whether these values still get written to the stackmap if there are no other live variables. So we should probably add a check here.

Reply:
A `debug_assert` is always a good thing in my book!

Reply:
Fixed in [358141e](https://github.com/ykjit/yk/pull/864/commits/358141e0ae1cf1d84faddccd16cd5463d5028895)